insert_deduplication_token setting for INSERT statement

devcrafter opened the pull request:
The setting allows a user to provide own deduplication semantic in MergeTree/ReplicatedMergeTree
If provided, it's used instead of data digest to generate block ID
So, for example, by providing a unique value for the setting in each INSERT statement,
user can avoid the same inserted data being deduplicated

Inserting data within the same INSERT statement are split into blocks
according to the *insert_block_size* settings
(max_insert_block_size, min_insert_block_size_rows, min_insert_block_size_bytes).
Each block with the same INSERT statement will get an ordinal number.
The ordinal number is added to insert_deduplication_token to get block dedup token
i.e. <token>_0, <token>_1, ... Deduplication is done per block
So, to guarantee deduplication for two same INSERT queries,
dedup token and number of blocks have to be the same

This closes: #7461

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
The setting allows a user to provide own deduplication semantic in MergeTree/ReplicatedMergeTree If provided, it's used instead of data digest to generate block ID. So, for example, by providing a unique value for the setting in each INSERT statement, the user can avoid the same inserted data being deduplicated. This closes: #7461


Detailed description / Documentation draft:

TODO

> By adding documentation, you'll allow users to try your new feature immediately, not when someone else will have time to document it later. Documentation is necessary for all features that affect user experience in any way. You can add brief documentation draft above, or add documentation right into your patch as Markdown files in [docs](https://github.com/ClickHouse/ClickHouse/tree/master/docs) folder.
> If you are doing this for the first time, it's recommended to read the lightweight [Contributing to ClickHouse Documentation](https://github.com/ClickHouse/ClickHouse/tree/master/docs/README.md) guide first.
> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

devcrafter committed:
insert_deduplication_token setting for INSERT statement

The setting allows a user to provide own deduplication semantic in Replicated*MergeTree
If provided, it's used instead of data digest to generate block ID
So, for example, by providing a unique value for the setting in each INSERT statement,
user can avoid the same inserted data being deduplicated

Inserting data within the same INSERT statement are split into blocks
according to the *insert_block_size* settings
(max_insert_block_size, min_insert_block_size_rows, min_insert_block_size_bytes).
Each block with the same INSERT statement will get an ordinal number.
The ordinal number is added to insert_deduplication_token to get block dedup token
i.e. <token>_0, <token>_1, ... Deduplication is done per block
So, to guarantee deduplication for two same INSERT queries,
dedup token and number of blocks to have to be the same

Issue: #7461

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=32304) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=32304) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=32304) it.</sub>

alexey-milovidov commented:
@mergifyio update

alexey-milovidov commented:
@devcrafter Could you please allow edits to your branch?

alexey-milovidov commented on src/Storages/MergeTree/IMergeTreeDataPart.h:
```diff
@@ -170,7 +170,8 @@ class IMergeTreeDataPart : public std::enable_shared_from_this<IMergeTreeDataPar
     /// Compute part block id for zero level part. Otherwise throws an exception.
-    String getZeroLevelPartBlockID() const;
+    /// If token is not empty, block id is calculated based on it instead of block data
+    String getZeroLevelPartBlockID(const String & token = String()) const;
```
Default value is unnecessary.

devcrafter replied:
Yes, but after deduplication token will be added to non-replicated MergeTree  as well

alexey-milovidov commented on src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp:
```diff
@@ -160,8 +160,16 @@ void ReplicatedMergeTreeSink::consume(Chunk chunk)
+            {
+                /// multiple blocks can be inserted within the same insert query
+                /// an ordinal number is added to dedup token to generate a distinctive block id for each block
+                block_dedup_token += fmt::format("_{}", chunk_dedup_seqnum++);
```
Let's split into two statements.

alexey-milovidov commented on src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp:
```diff
@@ -160,8 +160,16 @@ void ReplicatedMergeTreeSink::consume(Chunk chunk)
+                /// multiple blocks can be inserted within the same insert query
+                /// an ordinal number is added to dedup token to generate a distinctive block id for each block
+                block_dedup_token += fmt::format("_{}", chunk_dedup_seqnum++);
+                LOG_TRACE(log, "Deduplication token: {}", block_dedup_token);
```
Let's remove.

devcrafter commented:
> @devcrafter Could you please allow edits to your branch?

Done

devcrafter commented:
@alexey-milovidov I still want to complete the PR with dedup token for non-replicated MergeTree. Don't you mind assigning it to me?

alexey-milovidov commented:
Yes, sure. I expect both Replicated and non-Replicated can be implemented in this PR.

alesapin commented:
@mergifyio update

mergify[bot] committed:
Merge branch 'master' into deduplication_token_7461

devcrafter committed:
insert_deduplication_token support in non-replicated MergeTree

devcrafter committed:
Merge remote-tracking branch 'origin/master' into deduplication_token_7461

devcrafter committed:
Fix test for insert_deduplication_token
+ sync replica before checks in replica test
+ use different table names for multi block tests

alexey-milovidov committed:
Update 02124_insert_deduplication_token_multiple_blocks.sh

alexey-milovidov committed:
Update 02124_insert_deduplication_token_multiple_blocks_replica.sh

alexey-milovidov commented:
GitHub is notoriously unstable in the last day, so CI is delayed:
https://news.ycombinator.com/item?id=30109187
